src/MonitorProfiler/MainProfiler/ThreadData.h

```diff
@@ -0,0 +1,37 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include "corhlpr.h"
+#include "corprof.h"
+#include "../Logging/Logger.h"
+
+/// <summary>
+/// Class representing common data for a single thread.
+/// </summary>
+class ThreadData
+{
+public:
+    static const FunctionID NoFunctionId = 0;
+
+private:
+    FunctionID _exceptionCatcherFunctionId;
+    bool _hasException;
+    std::mutex _mutex;
```

Comment:
Do we actually need a per thread mutex? While the unordered map of threads certainly needs guarding, do we ever read or mutate the data cross thread boundaries?

Assuming we do, would it be possible to read/write _hasException and _exceptionCatcherFunctionId using interlocked functions to avoid locking on a per thread basis?

Assuming we still need the mutex, is the std::mutex implementation scalable across let's say thousands of threads?

Reply:
I've removed the mutex and did not add the debug same-thread enforcement.